File: CodeWars/Mathematics/DiophantineEquation/DiophantineEquation.rs

fn solequa(n: u64) -> Vec<(u64, u64)> {
    let mut result = vec![];

    for i in 1..(n as f64).sqrt().floor() as u64 + 1_u64 {
        // i == x-2y
        // j == x+2y
        if n%i == 0  {
            let j = n/i;

            if (j-i)%4 == 0 {
                // j-i == 4y
                // i+j == 2x
                result.push(((i+j)/2, (j-i)/4));
            }
        }
    }

    result
}

fn testing(n: u64, exp: Vec<(u64, u64)>) -> () {
    assert_eq!(solequa(n), exp)
}

fn main() {
    testing(5, vec![(3, 1)]);
    testing(20, vec![(6, 2)]); 
    testing(9001, vec![(4501, 2250)]);
    testing(9004, vec![(2252, 1125)]);
}

File: CodeWars/Mathematics/DiophantineEquation/DiophantineEquation.py
import unittest
from math import sqrt, floor


def sol_equa(n):
    result = []

    for i in range(1, int(floor(sqrt(n)))+1):
        if n % i == 0:
            j = n/i

            if (j-i) % 4 == 0:
                result.append([(i+j)/2, (j-i)/4])

    return result


class TestExercise(unittest.TestCase):
    def test_sol_equa(self):
        self.assertEqual(sol_equa(5), [[3, 1]])
        self.assertEqual(sol_equa(20), [[6, 2]])
        self.assertEqual(sol_equa(9001), [[4501, 2250]])
        self.assertEqual(sol_equa(9004), [[2252, 1125]])


if __name__ == '__main__':
    unittest.main()
